src/Features/Core/Portable/NavigateTo/NavigateToSearcher.cs

```diff
@@ -81,111 +104,187 @@ internal async Task SearchAsync()
 
-            // First, if there's an active document, search that project first, prioritizing
-            // that active document and all visible documents from it.
-            if (activeDocument != null)
+            // We weren't fully loaded *but* we reported some items to the user, then consider that
+            // good enough for now.  The user will have some results they can use, and (in the case
+            // that we actually examined the cache for data) we will tell the user that the results
+            // may be incomplete/inaccurate and they should try again soon.
+            if (itemsReported > 0)
             {
-                var activeProject = activeDocument.Project;
-                processedProjects.Add(activeProject);
+                return projectResults.All(t => t.location == NavigateToSearchLocation.Latest);
+            }
 
-                var visibleDocsFromProject = visibleDocs.Where(d => d.Project == activeProject);
-                var priorityDocs = ImmutableArray.Create(activeDocument).AddRange(visibleDocsFromProject);
+            // We didn't have any items reported *and* we weren't fully loaded.  If it turns out
+            // that some of our projects were using cached data then we can try searching them
+            // again, but this tell them to use the latest data.  The ensures the user at least gets
+            // some result instead of nothing.
+            var projectsUsingCache = projectResults.Where(t => t.location == NavigateToSearchLocation.Cache).SelectAsArray(t => t.project);
+            await ProcessProjectsAsync(ImmutableArray.Create(projectsUsingCache), isFullyLoaded: true, cancellationToken).ConfigureAwait(false);
 
-                // Search the active project first.  That way we can deliver results that are
-                // closer in scope to the user quicker without forcing them to do something like
-                // NavToInCurrentDoc
-                await Task.Run(() => SearchAsync(activeProject, priorityDocs, seenItems, isFullyLoaded), _cancellationToken).ConfigureAwait(false);
+            // we did a full uncached search.  return true to indicate that no message should be
+            // shown to the user.
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a sequence of groups of projects to process.  The sequence is in priority order,
+        /// and all projects in a particular group should be processed before the next group.  This
+        /// allows us to associate CPU resources in likely areas the user wants, while also still
+        /// allowing for good parallelization.
+        /// </summary>
+        private ImmutableArray<ImmutableArray<Project>> GetOrderedProjectsToProcess()
+        {
+            // If we're only searching the current doc, we don't need to examine anything else but that.
+            if (_searchCurrentDocument)
+            {
+                var project = _currentDocument?.Project;
```

Comment:
Technically no.  We were asked to search the current document, but it's possible we couldn't determine the active doc:

```
                var documentService = _solution.Workspace.Services.GetRequiredService<IDocumentTrackingService>();
                var activeId = documentService.TryGetActiveDocument();
                _currentDocument = activeId != null ? _solution.GetDocument(activeId) : null;
```

This could potentially happen, for example, if NavTo is called very quickly on launch before we've even connected a buffer to the roslyn workspace.

Reply:
(note: i discovered this because i tried to amke that optimization.  i'll doc as such).